src/pool/srv_pool.c

```diff
@@ -6777,3 +6811,88 @@ ds_pool_target_status_check(struct ds_pool *pool, uint32_t id, uint8_t matched_s
 
 	return target->ta_comp.co_status == matched_status ? 1 : 0;
 }
+
+int
+ds_pool_svc_load_map(struct ds_pool_svc *ds_svc, struct pool_map **map)
+{
+	struct pool_svc		*svc = pool_ds2svc(ds_svc);
+	struct rdb_tx		 tx = { 0 };
+	int			 rc;
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc == 0) {
+		ABT_rwlock_rdlock(svc->ps_lock);
+		rc = read_map(&tx, &svc->ps_root, map);
+		ABT_rwlock_unlock(svc->ps_lock);
+		rdb_tx_end(&tx);
+	}
+
+	if (rc != 0)
+		D_ERROR("Failed to load pool map for pool "DF_UUIDF": "DF_RC"\n",
+			DP_UUID(svc->ps_uuid), DP_RC(rc));
+
+	return rc;
+}
+
+int
+ds_pool_svc_flush_map(struct ds_pool_svc *ds_svc, struct pool_map *map, uint32_t version)
+{
+	struct pool_svc		*svc = pool_ds2svc(ds_svc);
+	struct pool_buf		*buf = NULL;
+	struct rdb_tx		 tx = { 0 };
+	int			 rc = 0;
+
+	rc = rdb_tx_begin(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term, &tx);
+	if (rc != 0) {
+		D_ERROR("Failed to begin TX for updating pool "DF_UUIDF" map with version %u: "
+			DF_RC"\n", DP_UUID(svc->ps_uuid), version, DP_RC(rc));
+		goto out;
+	}
+
+	ABT_rwlock_wrlock(svc->ps_lock);
+
+	rc = pool_buf_extract(map, &buf);
+	if (rc != 0) {
+		D_ERROR("Failed to extract buf for updating pool "DF_UUIDF" map with version %u: "
+			DF_RC"\n", DP_UUID(svc->ps_uuid), version, DP_RC(rc));
+		goto out_lock;
+	}
+
+	rc = write_map_buf(&tx, &svc->ps_root, buf, version);
+	if (rc != 0) {
+		D_ERROR("Failed to write buf for updating pool "DF_UUIDF" map with version %u: "
+			DF_RC"\n", DP_UUID(svc->ps_uuid), version, DP_RC(rc));
+		goto out_buf;
+	}
+
+	rc = rdb_tx_commit(&tx);
+	if (rc != 0) {
+		D_ERROR("Failed to commit TX for updating pool "DF_UUIDF" map with version %u: "
+			DF_RC"\n", DP_UUID(svc->ps_uuid), version, DP_RC(rc));
+		goto out_buf;
+	}
+
+	/* Update svc->ps_pool to match the new pool map. */
+	rc = ds_pool_tgt_map_update(svc->ps_pool, buf, version);
+	if (rc != 0) {
+		D_ERROR("Failed to refresh local pool "DF_UUIDF" map with version %u: "
+			DF_RC"\n", DP_UUID(svc->ps_uuid), version, DP_RC(rc));
+		/*
+		 * Have toresign to avoid handling future requests with stale pool map cache.
+		 * Continue to distribute the new pool map to other pool shards since the RDB
+		 * has already been updated.
+		 */
+		rdb_resign(svc->ps_rsvc.s_db, svc->ps_rsvc.s_term);
```

Comment:
Failing to update the local map implies that the local secondary group for this pool may not have the latest membership. In this state, it's simpler to just give up, instead of continuing to take some chance. When a new leader steps up, it will distribute the new pool map, schedule rebuild jobs, (and in the future also do what replace_failed_replicas does).

Reply:
Under DAOS check mode, in spite of PS leader itself is down or some other is down during checking the pool membership, then the DAOS check for this pool will be marked as aborted.
On the other hand, if the PS leader switches to other engine because of short time network split without engine down, then it will not cause DAOS check to be failed even if both the old PS leader and the new PS leader do DAOS pool membership check in parallel. It may cause some redundant check, but not error.
So here, the current PS leader must has the latest membership. It has already updated the pool map in RDB, but failed to refresh the pool map (attached to the pool instance) in-DRAM. If we give up, means the DAOS check for this pool will be failed, but if we try to distributed the update to other engines, we may have chance to continue the DAOS check.

Reply:
The rest of the PS code assumes that the local secondary group _always_ reflects the latest pool map. At least, I'd suggest skipping the following the map_dist and reconfigure calls in this case to avoid adding burden to non-chk code. Also, do you really intend to return the nonzero rc to chk like this patch does?

Reply:
OK, if do not have others to change, I prefer to adjust it in the subsequent patch https://github.com/daos-stack/daos/pull/9867 which will be rebased after landing this one.

As for the return value, I think it is better to return it the the caller, it is caller's duty to determine the next step. For CHK case, if not specify "failout",  then will go ahead.

How do you think?

Reply:
OK, fine with me.